### PR TITLE
Media endpoint DataViews and merged datasets support

### PR DESCRIPTION
**Pull request by FrankApiyo**
### Changes / Features implemented
- Ensure attachments endpoint works as expected for Merged datasets and filtered datasets
### Steps taken to verify this change does what is intended
- [x] Added test cases
- [x] QA (todo)
### Side effects of implementing this change
N/A

**Before submitting this PR for review, please make sure you have:**

  - [x] Included tests
  - [x] Updated documentation

closes https://github.com/onaio/onadata/issues/2375
Closes https://github.com/onaio/zebra/issues/7352

### CODE REVIEW

File: onadata/apps/api/tests/viewsets/test_dataview_viewset.py [hunk 1]
```diff
@@ -86,7 +87,7 @@ def test_filter_to_field_lookup(self):
 
     def test_get_field_lookup(self):
         self.assertEqual(
-            get_field_lookup("q1", "="), "json__q1__iexact"
```
Comment (DavisRayM) [inline onadata/apps/api/tests/viewsets/test_dataview_viewset.py:89]: Why are we removing the case insensitive search ?

File: onadata/apps/api/viewsets/dataview_viewset.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def filter_to_field_lookup(filter_string):
     Converts a =, < or > to a django field lookup
     """
     if filter_string == "=":
-        return "__iexact"
```
Comment (DavisRayM) [inline onadata/apps/api/viewsets/dataview_viewset.py:67]: Same question

File: onadata/apps/api/viewsets/dataview_viewset.py [hunk 3]
```diff
@@ -77,25 +77,47 @@ def get_field_lookup(column, filter_string):
     return "json__" + column + filter_to_field_lookup(filter_string)
 
 
-def apply_filters(instance_qs, filters):
+def get_filter_kwargs(filters):
```
Comment (DavisRayM) [inline onadata/apps/api/viewsets/dataview_viewset.py:80]: I'm a bit against changing function signatures; Not sure of the potential impact of doing so

**Review by DavisRayM — CHANGES_REQUESTED** [review]
Please add more details on the reason for the changes and what they achieve

**Comment by FrankApiyo** [comment]
> Please add more details on the reason for the changes and what they achieve

I opened an issue [here](https://github.com/onaio/onadata/issues/2375) with a clearer description of what this changes fix

File: onadata/apps/api/viewsets/dataview_viewset.py (hunk 3 above)
Comment (FrankApiyo) [reply onadata/apps/api/viewsets/dataview_viewset.py:80]: There's only one place the `apply_filters` function is used (I have ensured it's updated to work with the new signature)

This change reduced the number of things `apply_filters` does (Gives it a single responsibility)

File: onadata/apps/api/tests/viewsets/test_dataview_viewset.py (hunk 1 above)
Comment (FrankApiyo) [reply onadata/apps/api/tests/viewsets/test_dataview_viewset.py]: seems there was no need for this change, reverted!

File: onadata/apps/api/viewsets/dataview_viewset.py (hunk 3 above)
Comment (FrankApiyo) [reply onadata/apps/api/viewsets/dataview_viewset.py:80]: I have also made sure that these changes are well tested

File: onadata/apps/api/viewsets/dataview_viewset.py [hunk 4]
```diff
@@ -77,25 +77,47 @@ def get_field_lookup(column, filter_string):
...
+
+def get_dataview_instances(dataview):
+    """
+    Get all instances that belong to ths dataview
+    """
+    return apply_filters(
+        dataview.xform.instances.filter(
+            deleted_at__isnull=True
+        ), dataview.query
+    ).order_by('id')
```
Comment (DavisRayM) [inline onadata/apps/api/viewsets/dataview_viewset.py:117]: Do we have to do the ordering ? Can we just let the instances be returned in the same order stored in the database

Comment (FrankApiyo) [reply onadata/apps/api/viewsets/dataview_viewset.py:117]: No, this isn't absolutely necessary; I think the instances are already sorted by ID in the DB

Comment (FrankApiyo) [reply onadata/apps/api/viewsets/dataview_viewset.py:117]: Removing

File: onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py [hunk 5]
```diff
@@ -540,6 +552,46 @@ def test_get_osm_data_kwargs(self):
...
+        del response_list[0]['url']
+        del response_list[0]['xform']
+        del response_list[1]['small_download_url']
+        del response_list[1]['medium_download_url']
+        del response_list[1]['download_url']
+        del response_list[1]['filename']
+        del response_list[1]['id']
+        del response_list[1]['instance']
+        del response_list[1]['url']
+        del response_list[1]['xform']
```
Comment (DavisRayM) [inline onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py:587]:
```suggestion
        del response_list[0]['small_download_url']
        del response_list[0]['medium_download_url']
        del response_list[0]['download_url']
        del response_list[0]['filename']
        del response_list[0]['id']
        del response_list[0]['instance']
        del response_list[0]['url']
        del response_list[0]['xform']
        del response_list[1]['small_download_url']
        del response_list[1]['medium_download_url']
        del response_list[1]['download_url']
        del response_list[1]['filename']
        del response_list[1]['id']
        del response_list[1]['instance']
        del response_list[1]['url']
        del response_list[1]['xform']
```
Might just be nitpicking here but these seem like things that can be determined from the stored attachment object to me...

**Review by DavisRayM — APPROVED** [review]

File: onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py (hunk 5 above)
Comment (FrankApiyo) [reply onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py]: I have refactored the test case, have a look!

File: onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py [hunk 6]
```diff
@@ -540,6 +559,32 @@ def test_get_osm_data_kwargs(self):
...
+        all_instances = Instance.objects.filter(xform__in=merged_xform.xforms.all())
+        for instance in all_instances:
+            _add_attachments_to_instances(instance)
+        request = self.factory.get(
+            "/?merged_xform=" + str(merged_xform.pk),
+            **self.extra)
+        response = attachment_list_view(request)
+        response_list = json.loads(json.dumps(response.data))
+        serialized_attachments = json.loads(json.dumps(
+            AttachmentSerializer(
```
Comment (ukanga) [inline onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py:579]: @FrankApiyo Is this not just being repetitive? Is a comparison of `response.data` and `serializer.data` not possible or equivalent?

Comment (FrankApiyo) [reply onadata/apps/api/tests/viewsets/test_merged_xform_viewset.py]: Makes sense, I have refactored; now we compare `response.data` with `serializer.data`

**Review by ukanga — APPROVED** [review]

**Comment by KipSigei** [comment]
@FrankApiyo Could we update the aAPI documentation for this change?

**Review by KipSigei — APPROVED** [review]

**Review by KipSigei — APPROVED** [review]

**Pull request merged**